File: Notes/5-2_CreateBiTree_sequence.c
// This is the C implementation of 5-2_CreateBiTree_sequence.py
/* An algorithm to generete an unique binary tree with its preorder and
inorder sequence.
*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
typedef struct treeNode *Tree;
struct treeNode
{
    int value;
    Tree left;
    Tree right;
};

Tree CreateBiTree(char *preseq, char *inseq);
void PostOrderTraversal(Tree root);
Tree CreateBiTreeNode(int value);

int main(void)
{
    char *preseq = "ABCDEFGHI";
    char *inseq = "BCAEDGHFI";
    // Tree bitree = CreateBiTree(preseq, inseq);
    // PostOrderTraversal(bitree);
    printf("Ha, it is a boring test.\n");
    printf("Length: %d\n", (int)strlen(inseq));
    char cha = *(strchr(preseq, 'E'));
    char first = *preseq;
    printf("The char is %c", cha);
    printf("First char of preseq: %c\n", first);
    CreateBiTreeNode(first);
    return 0;
}

Tree CreateBiTree(char *preseq, char *inseq)
{
    char curr;
    if (strlen(preseq) <= 0 && strlen(inseq) <= 0)
        return NULL;
    else
    {
        curr = *preseq;
        // TODO:
        // split inorder sequence into two parts
        // split preorder sequence into two parts
        // NOT DONE YET
        return NULL;
    }
}

void PostOrderTraversal(Tree root)
{
    if (root)
    {
        PostOrderTraversal(root->left);
        PostOrderTraversal(root->right);
        printf("%d ", root->value);
    }
}

Tree CreateBiTreeNode(int value)
{
    Tree node = (Tree)malloc(sizeof(struct treeNode));
    node->value = value;
    node->left = node->right = NULL;
    return node;
}

File: Notes/5-2_CreateBiTree_sequence.py
'''
Excercise 5.11.5.py
An algorithm to generate an unique binary tree with its preorder
and inorder sequence.
Frequently used helper function when learning binary tree data
structure. And is simpler than file 5-2_CreateBiTree_recursive.py
'''

class treeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def CreateBiTree(preseq, inseq):
    '''
    Input is two string or list of tree traversal sequences, inorder and
    preorder. This function constructs the unique tree that should generate
    the inputed sequences.
    '''
    if not (preseq or inseq):
        return None
    else:
        curr = preseq[0]
        in_left_part, in_right_part = inseq.split(curr)
        pre_left_part = preseq[1: len(in_left_part)+1]
        pre_right_part = preseq[len(in_left_part)+1: ]
        root = treeNode(curr)
        root.left = CreateBiTree(pre_left_part, in_left_part)
        root.right = CreateBiTree(pre_right_part, pre_right_part)
        return root

def traversal(tree):
    stack = []
    curr = tree
    while curr or stack:
        while curr:
            print(curr.val, end=" ")
            stack.append(curr)
            curr = curr.left
        if not curr:
            curr = stack.pop()
            curr = curr.right

if __name__ == "__main__":
    preorder = "ABCDEFGHI"
    inorder = "BCAEDGHFI"
    tree = CreateBiTree(preorder, inorder)
    traversal(tree)
    